example/src/Client/utils/AccountsManager.ts

```diff
@@ -152,43 +203,122 @@ export class AccountsManager {
+        account.key.viewKey,
+        BigInt(position),
+      );
+
+      account.assetIdByNoteHash.set(deserializedNote.hash(), assetId);
+      account.noteHashByNullifier.set(nullifier, deserializedNote.hash());
 
       // Register note
-      assetEntry.decryptedNotes.push({
+      assetContent.set(deserializedNote.hash(), {
         accountId: publicKey,
-        note: foundNode,
+        note: deserializedNote,
         spent: false,
         transactionHash: tx.hash,
         index,
-        nullifier: null, // @todo: Get nullifier
+        nullifier,
         blockHash: block.hash,
         sequence: block.sequence,
       });
 
-      // Update balance
-      const currentBalance = account.assets.get(assetId)?.balance ?? BigInt(0);
-      assetEntry.balance = currentBalance + amount;
-
       logThrottled(
-        `Account ${publicKey} has ${assetEntry.decryptedNotes.length} notes for asset ${assetId}`,
+        `Account ${publicKey} has ${
+          assetContent.size
+        } notes for asset ${assetId.toString("hex")}`,
         10,
-        assetEntry.decryptedNotes.length,
+        assetContent.size,
       );
     }
   }
+
+  private _processSpend(spend: LightSpend) {
+    for (const account of this.accounts.values()) {
+      const noteHash = account.noteHashByNullifier.get(spend.nf);
+
+      if (!noteHash) return;
+
+      const assetId = account.assetIdByNoteHash.get(noteHash);
+
+      if (!assetId) return;
+
+      const note = account.assets.get(assetId)?.get(noteHash);
+
+      if (!note) return;
+
+      note.spent = true;
+    }
+  }
+
+  private _handleBlockReorg(block: LightBlock) {
+    // To process a reorg for a block, we need to undo what spends and outputs normally do.
+    // For spends, we need to mark the note as unspent and update the appropriate mappings.
+    // For outputs, we need to remove the note.
+
+    // We have to process reorg'd blocks for each account
+    for (const [_, account] of this.accounts) {
+      block.transactions.forEach((tx) => {
+        // First we'll process spends
+        tx.spends.forEach((spend) => {
+          // If we have a note hash for the nullifier, it means we have a note for this account.
+          const noteHash = account.noteHashByNullifier.get(spend.nf);
+
+          if (!noteHash) return;
+
+          // If we found a note hash, the following should exist.
+          const assetId = account.assetIdByNoteHash.get(noteHash);
+          const assetContent = assetId && account.assets.get(assetId);
+          const storedNote = assetContent && assetContent.get(noteHash);
+
+          if (!assetId || !assetContent || !storedNote) return;
+
+          // Since we're undoing a spend, we need to mark the note as unspent.
+          storedNote.spent = false;
+        });
+
+        // Next we'll process outputs
+        tx.outputs.forEach((output) => {
+          const note = new NoteEncrypted(output.note);
+
+          // Decrypt note using view key
```

Comment:
I think deleting the notes for a given block is also an effective mechanism, which would require less computation. Shipping it this way seems fine though.

Reply:
I'm sold on storing notes by block, but yeah, let's do it as a separate PR.